File: codeforces/round379/D.cc
// Codeforces Round #379 (Div. 2)
// http://codeforces.com/contest/734

// ID: 734D (Anton and Chess)

#include <stdio.h>
#include <vector>
#include <algorithm>
using namespace std;

struct Coord { char type; int x, y; };
	Coord king;	
	vector<Coord> blacks, attackers;

bool lessX(Coord& a, Coord& b) { return a.x < b.x; }
bool lessY(Coord& a, Coord& b) { return a.y < b.y; }
bool greaterX(Coord& a, Coord& b) { return a.x > b.x; }
bool greaterY(Coord& a, Coord& b) { return a.y > b.y; }
using Sort = bool (*)(Coord&, Coord&);

void choose(int dx, int dy, Sort sortFn) {
	vector<Coord> candidates;
	// filter
	for(int i=0; i<blacks.size(); i++){
		int diffX = blacks[i].x - king.x;
		int diffY = blacks[i].y - king.y;
		bool ok = false;
		if(dx == 0) ok = diffX == 0 && diffY / dy >= 1;
		else if(dy == 0) ok = diffY == 0 && diffX / dx >= 1;
		else{
			int ry = diffY / dy, rx = diffX / dx;
			ok = ry >= 1 && rx >= 1 && ry == rx;
		}
		if(ok) candidates.push_back(blacks[i]);
	}
	if(candidates.size() == 0) return;
	// sort
	sort(candidates.begin(), candidates.end(), sortFn);
	attackers.push_back(candidates[0]);
}

bool checkmate(Coord& black, Coord& king) {
	int dx = black.x - king.x;
	int dy = black.y - king.y;
	if(black.type == 'R' || black.type == 'Q'){
		if(dx == 0 || dy == 0) return true;
	}
	if(black.type == 'B' || black.type == 'Q'){
		if(dy != 0 && abs(dx/dy) == 1) return true;
	}
	return false;
}

void solve() {
	// input
	int n;
	scanf("%d", &n);
	scanf("%d %d", &king.x, &king.y);
	for(int i=0; i<n; i++){
		Coord black;
		char buf[3];
		scanf("%s %d %d", buf, &black.x, &black.y);
		black.type = buf[0];
		blacks.push_back(black);
	}

	// select only nearest attackers for 8 directions
	choose(-1, -1, greaterX);  choose(0, -1, greaterY);  choose(1, -1, lessX);
	choose(-1, 0, greaterX);                             choose(1, 0, lessX);
	choose(-1, 1, greaterX);   choose(0, 1, lessY);      choose(1, 1, lessX);

	// check
	for(int i=0; i<attackers.size(); i++){
		if(checkmate(attackers[i], king)){
			puts("YES");
			return;
		}
	}
	puts("NO");
}

int main() {
	solve();
	return 0;
}

File: codeforces/round379/D.hs
-- Codeforces Round #379 (Div. 2)
-- http://codeforces.com/contest/734

-- ID: 734D (Anton and Chess)
import Control.Monad
import Data.List
import Data.Int
import Data.Maybe
import qualified Data.ByteString as BS
import qualified Data.ByteString.Char8 as BS8

type Piece = (String, Int, Int)

getInts = (map (fst . fromJust . BS8.readInt) . BS8.words) `fmap` BS.getLine :: IO [Int]
parseInt = fst . fromJust . BS8.readInt

lessX (_,x1,_) (_,x2,_) = compare x1 x2
lessY (_,_,y1) (_,_,y2) = compare y1 y2

main = do
	n <- read `fmap` getLine :: IO Int
	[kingX, kingY] <- (map read . words) `fmap` getLine :: IO [Int]
	blacks <- replicateM n $ do
		[post, x, y] <- BS8.words `fmap` BS.getLine
		return (BS8.unpack post, parseInt x, parseInt y)
	let king = ("K", kingX, kingY)
	let dirs = [(-1,-1,flip lessX), (0,-1,flip lessY), (1,-1,lessX),
				(-1,0,flip lessX), (1,0,lessX),
				(-1,1,flip lessX), (0,1,lessY), (1,1,lessX)]
	let attackers = map fromJust $ filter isJust $ map (choose king blacks) dirs
	if any (checkmate king) attackers
	then putStrLn "YES"
	else putStrLn "NO"

choose king blacks (dx, dy, sorter) = case (sortBy sorter . filter (attack dx dy king)) blacks of
	[] -> Nothing
	xs -> Just (head xs)
attack dx dy (_, kx, ky) (_, x, y)
	| dx == 0 = diffX == 0 && ry >= 1
	| dy == 0 = diffY == 0 && rx >= 1
	| otherwise = rx >= 1 && rx >= 1 && ry == rx
	where
		diffX = x - kx
		diffY = y - ky
		rx = diffX `div` dx
		ry = diffY `div` dy

checkmate (_, kx, ky) (post, x, y)
	| post == "R" = dx == 0 || dy == 0
	| post == "B" = dy /= 0 && (abs (dx `div` dy) == 1)
	| post == "Q" = True
	where
		dx = x - kx
		dy = y - ky

